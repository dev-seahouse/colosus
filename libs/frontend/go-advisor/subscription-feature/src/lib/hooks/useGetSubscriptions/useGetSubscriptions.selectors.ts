import { useQuery } from '@tanstack/react-query';
import { DateTime } from 'luxon';

import { useCallback } from 'react';
import { convertUnixToDate } from './useGetSubscriptions.utils';
import { getSubscriptionsQuery } from './useGetSubscriptions';
import type { GetSubscriptionsData } from './useGetSubscriptions';
import getSubscriptionAmount from '../../utils/getSubscriptionAmount/getSubscriptionAmount';

/**
 * hoook to return available subscription
 */
export const useSelectSubscriptionQuery = () => {
  return useQuery({
    ...getSubscriptionsQuery(),
    select: useCallback((data: GetSubscriptionsData) => {
      const subscription = data.subscriptions[0];
      const { currency, current_period_end, current_period_start, items } =
        subscription;

      return {
        currency,
        subscriptionPlan: items.data[0].plan.nickname,
        subscriptionFee: getSubscriptionAmount(
          items.data[0].plan.amount_decimal
        ),
        subscriptionInterval: items.data[0].plan.interval,
        startBillingDate: convertUnixToDate(current_period_start),
        nextBillingDate: convertUnixToDate(
          DateTime.fromSeconds(current_period_end).plus({ day: 1 }).toSeconds()
        ),
      };
    }, []),
  });
};
