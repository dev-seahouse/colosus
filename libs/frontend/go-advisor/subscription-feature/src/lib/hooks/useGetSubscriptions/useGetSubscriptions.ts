import { useQuery } from '@tanstack/react-query';
import { StripeIntegrationAdvisorSubscriptionsApi } from '@bambu/api-client';
import type { StripeIntegrationGetSubscriptionsResponseDto } from '@bambu/api-client';
import type { QueryClient } from '@tanstack/react-query';

export type GetSubscriptionsData = StripeIntegrationGetSubscriptionsResponseDto;

export const getSubscriptionsQuery = () => ({
  queryKey: ['getSubscriptions'],
  queryFn: async () => {
    const stripeIntegrationAdvisorSubscriptionsApi =
      new StripeIntegrationAdvisorSubscriptionsApi();

    const res =
      await stripeIntegrationAdvisorSubscriptionsApi.getSubscriptions();

    return res.data;
  },
});

export const getSubscriptionsLoader =
  (queryClient: QueryClient) => async (): Promise<GetSubscriptionsData> => {
    const query = getSubscriptionsQuery();

    // ⬇️ return data or fetch it
    return (
      queryClient.getQueryData(query.queryKey) ??
      (await queryClient.fetchQuery(query))
    );
  };

export interface UseGetSubscriptionsOptions {
  initialData?: Partial<GetSubscriptionsData>;
}
export function useGetSubscriptions({
  initialData,
}: UseGetSubscriptionsOptions = {}) {
  return useQuery({
    ...getSubscriptionsQuery(),
    initialData,
  });
}

export default useGetSubscriptions;
