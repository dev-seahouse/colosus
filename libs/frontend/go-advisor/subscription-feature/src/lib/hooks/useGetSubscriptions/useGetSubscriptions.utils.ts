import { DateTime } from 'luxon';

export const convertUnixToDate = (unix: number, format = 'd LLLL yyyy') => {
  return DateTime.fromSeconds(unix).toFormat(format);
};
