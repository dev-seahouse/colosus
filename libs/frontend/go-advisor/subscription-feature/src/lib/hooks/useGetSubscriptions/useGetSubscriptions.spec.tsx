import {
  renderHook,
  queryClientWrapper,
  waitFor,
} from '@bambu/react-test-utils';

import useGetSubscriptions from './useGetSubscriptions';

describe('useGetPrices', () => {
  it('should render successfully', async () => {
    const { result } = renderHook(() => useGetSubscriptions(), {
      wrapper: queryClientWrapper,
    });

    await waitFor(() => expect(result.current.isSuccess).toBe(true));

    expect(result.current.data).toMatchSnapshot();
  });
});
