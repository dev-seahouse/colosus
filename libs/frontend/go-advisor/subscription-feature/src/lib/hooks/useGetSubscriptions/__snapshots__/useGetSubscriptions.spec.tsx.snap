// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useGetPrices > should render successfully 1`] = `
{
  "subscriptions": [
    {
      "application": null,
      "application_fee_percent": null,
      "automatic_tax": {
        "enabled": false,
      },
      "billing_cycle_anchor": 1681273057,
      "billing_thresholds": null,
      "cancel_at": null,
      "cancel_at_period_end": false,
      "canceled_at": null,
      "cancellation_details": {
        "comment": null,
        "feedback": null,
        "reason": null,
      },
      "collection_method": "charge_automatically",
      "created": 1681273057,
      "currency": "myr",
      "current_period_end": 1683865057,
      "current_period_start": 1681273057,
      "customer": "cus_NhJlr0T3WH1vsd",
      "days_until_due": null,
      "default_payment_method": "pm_1Mvv7kFabpOjRn7HpDkDMI2d",
      "default_source": null,
      "default_tax_rates": [],
      "description": null,
      "discount": null,
      "ended_at": null,
      "id": "sub_1Mvv7lFabpOjRn7HMNhMYiGf",
      "items": {
        "data": [
          {
            "billing_thresholds": null,
            "created": 1681273058,
            "id": "si_NhJlx6OkheunGI",
            "metadata": {},
            "object": "subscription_item",
            "plan": {
              "active": true,
              "aggregate_usage": null,
              "amount": 2000,
              "amount_decimal": "2000",
              "billing_scheme": "per_unit",
              "created": 1679897237,
              "currency": "myr",
              "id": "price_1Mq9D7FabpOjRn7HWILE9Cro",
              "interval": "month",
              "interval_count": 1,
              "livemode": false,
              "metadata": {},
              "nickname": "Connect",
              "object": "plan",
              "product": "prod_NbLut6uJw85qmi",
              "tiers_mode": null,
              "transform_usage": null,
              "trial_period_days": null,
              "usage_type": "licensed",
            },
            "price": {
              "active": true,
              "billing_scheme": "per_unit",
              "created": 1679897237,
              "currency": "myr",
              "custom_unit_amount": null,
              "id": "price_1Mq9D7FabpOjRn7HWILE9Cro",
              "livemode": false,
              "lookup_key": null,
              "metadata": {},
              "nickname": "Connect",
              "object": "price",
              "product": "prod_NbLut6uJw85qmi",
              "recurring": {
                "aggregate_usage": null,
                "interval": "month",
                "interval_count": 1,
                "trial_period_days": null,
                "usage_type": "licensed",
              },
              "tax_behavior": "unspecified",
              "tiers_mode": null,
              "transform_quantity": null,
              "type": "recurring",
              "unit_amount": 2000,
              "unit_amount_decimal": "2000",
            },
            "quantity": 1,
            "subscription": "sub_1Mvv7lFabpOjRn7HMNhMYiGf",
            "tax_rates": [],
          },
        ],
        "has_more": false,
        "object": "list",
        "url": "/v1/subscription_items?subscription=sub_1Mvv7lFabpOjRn7HMNhMYiGf",
      },
      "latest_invoice": "in_1Mvv7lFabpOjRn7HFfMfy19X",
      "livemode": false,
      "metadata": {},
      "next_pending_invoice_item_invoice": null,
      "object": "subscription",
      "on_behalf_of": null,
      "pause_collection": null,
      "payment_settings": {
        "payment_method_options": null,
        "payment_method_types": null,
        "save_default_payment_method": "off",
      },
      "pending_invoice_item_interval": null,
      "pending_setup_intent": null,
      "pending_update": null,
      "schedule": null,
      "start_date": 1681273057,
      "status": "active",
      "test_clock": null,
      "transfer_data": null,
      "trial_end": null,
      "trial_settings": {
        "end_behavior": {
          "missing_payment_method": "create_invoice",
        },
      },
      "trial_start": null,
    },
  ],
}
`;
